gluten-core/src/main/java/org/apache/gluten/memory/memtarget/spark/DynamicOffHeapSizingPolicyChecker.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.gluten.memory.memtarget.spark;
+
+import org.apache.gluten.GlutenConfig;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.atomic.AtomicLong;
+
+public final class DynamicOffHeapSizingPolicyChecker {
+  private static final Logger LOG =
+      LoggerFactory.getLogger(DynamicOffHeapSizingPolicyChecker.class);
+  private final long maxOnHeapMemoryInBytes = GlutenConfig.getConf().onHeapMemorySize();
+  private final AtomicLong usedOffHeapBytes = new AtomicLong();
+
+  DynamicOffHeapSizingPolicyChecker() {}
+
+  public boolean canBorrow(long size) {
+    if (size == 0) {
+      return true;
+    }
+
+    long totalMemory = Runtime.getRuntime().totalMemory();
+    long freeMemory = Runtime.getRuntime().freeMemory();
+    long usedOnHeapBytes = (totalMemory - freeMemory);
```

Comment:
Is this calculation safe enough? Say if total memory is 10G and free memory is 5G, then the JVM could still occupy 10G RAM from host machine, isn't it?

Reply:
As per https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/Runtime.html, totalMemory is the "the total amount of memory currently available for current and future objects, measured in bytes", i.e., the on-heap size set for Spark via spark.executor.memory in our case. freeMemory is the "an approximation to the total amount of memory currently available for future allocated objects", i.e., memory NOT currently in use by Java objects. 

In this calculation if the on-heap was set to 10GB and 5GB is unused by any Java objects, usedOnHeapBytes would give us 5GB. The JVM memory footprint in this case is 10GB. With the calculations above, we would allow up to 5GB of memory to be used for off-heap.

Does that help?

Reply:
To be transparent, however, you are 100% correct that this is not a strict / perfect quota. For example, if allocations happen on Java code, the logic above doesn't kick in, so we may oversubscribe memory usage.

Reply:
Thank for the explanation.

> In this calculation if the on-heap was set to 10GB and 5GB is unused by any Java objects, usedOnHeapBytes would give us 5GB. The JVM memory footprint in this case is 10GB. With the calculations above, we would allow up to 5GB of memory to be used for off-heap.

So in this case, the important one is that whether the total RSS of the JVM is calculated by which one of the following:

1. 10G total-heap + 5G off-heap = 15G
2. 5G used-heap + 5G off-heap = 10G

The code seems to map to the second calculation. But I am concerned that probably the first one is the real situation. If user configured Gluten/Spark with a 10G memory limit but the actual RSS goes 15G, that is unexpected.

Reply:
> To be transparent, however, you are 100% correct that this is not a strict / perfect quota. For example, if allocations happen on Java code, the logic above doesn't kick in, so we may oversubscribe memory usage.

Yes, but this one is another issue that doesn't seem to be easily solved in terms of the PR's implementation. Maybe we can temporarily just document this to let user know the risk.

Reply:
I see. Yes, the amount of memory that we are calculating as "committed but unused by the JVM", with the freeMemory API call, likely would be counted by the OSS as part of the RSS since the JVM likely touches those pages as they are added to the JVM heap free list. 

However, the counterpoint is that these pages will not see further memory access unless they are actually put to use, so if we consider them against the overall "used" quota, we are essentially leaving unused memory capacity at the table.

Perhaps a good way to approach this is to add a specific config with the proposed change that optionally allows for this more aggressive use of memory. If the config is off, say, the default, we will consider totalMemory as the total used memory, not discounting the freeMemory API value. If the setting is off, we will discount the freeMemory API value as the proposed changes stand now.

Thoughts?